Fix issue #19216 : Improving saving flow for preferences page using batch update

AkashPaloju opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes  #19216 .
2. This PR does the following: 
- Changes the current Saving flow in preferences page, for updating multiple preferences in a single request.
-  Adds a 'sticky-save' button in preferences page.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->
[sticky-save-desktop.webm](https://github.com/oppia/oppia/assets/119884665/998200e9-a40b-4e21-8fc3-931b2ea9dd32)

[latest-functionality.webm](https://github.com/oppia/oppia/assets/119884665/c6f2bb02-3290-420e-a59d-9761d1d69a74)

[latest-image-preview.webm](https://github.com/oppia/oppia/assets/119884665/7c74440b-887a-4e35-9d2d-b0625f521eba)

Below popup is shown if user doesn't click save after changing any value before leaving the page.

Web
[popup.webm](https://github.com/oppia/oppia/assets/119884665/535a98ad-d174-4896-8f63-a4139bfb6640)

mobile-popup

[latest-mob-popup.webm](https://github.com/oppia/oppia/assets/119884665/83d85c9e-8344-445e-82e8-8f76944c1397)


#### Proof of changes on desktop with slow/throttled network
[sticky-save-throttled.webm](https://github.com/oppia/oppia/assets/119884665/f44e4cfa-eb02-4e53-8fb7-20c9906d8dbe)

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
[SS-mobile.webm](https://github.com/oppia/oppia/assets/119884665/eda1fc7c-b92a-4567-97fb-046c12fe0c52)

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language
![SS-arabic](https://github.com/oppia/oppia/assets/119884665/a3c6d9ea-4286-4186-a421-42a9ee6d98a3)

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

AkashPaloju committed:
Changed update flow for updating multiple preferences in a single request

AkashPaloju committed:
Added i18n for SaveChanges, Removed unnecessary text

seanlip commented:
@AkashPaloju Please fix the failing tests.

AkashPaloju commented:
hi, @seanlip . Actually ,oppia-bot automatically requested code owner review when PR is opened.
One more thing is I am facing a glitch in Github, which I am trying to solve now. Though I have used git cli for making 2nd commit which fixes those failing tests. and changes are committed to the origin branch they are not reflecting in PR here.
![image](https://github.com/oppia/oppia/assets/119884665/d0b9d80e-65e7-45de-9bdb-eb893b4f6d80)

![image](https://github.com/oppia/oppia/assets/119884665/20337906-b78c-42b5-bff5-8e9c88bdd5a0)

Thanks!

AkashPaloju committed:
Changed a comment for better understanding

seanlip commented:
@AkashPaloju  What is the description of the commit in question? I see the commits in your screenshot mentioned in the Conversation tab here.

AkashPaloju committed:
Added tests for the new code

AkashPaloju commented:
> @AkashPaloju What is the description of the commit in question? I see the commits in your screenshot mentioned in the Conversation tab here

I have fixed that issue.  The 2nd commit was not visible here in Conversation Tab until I have made another commit. (see the below image, generally all commits undergo CI checks but this commit didn't, there is no failed or passed symbol near it.) 
![image](https://github.com/oppia/oppia/assets/119884665/fe74bd6b-1911-472c-b115-7a7c3224a6d8)

Now all failing checks are fixed. 
I have written i18n only for 'Save Changes' in en.json file for English.  PTAL
Thanks!

AkashPaloju committed:
Fixed imports issues

AkashPaloju committed:
Fix flaky Context service issue

AkashPaloju committed:
Changes in e2e tests

AkashPaloju commented:
@seanlip Could you please rerun this tests. As of now all tests are passed except the following e2e test due to click interceptions.
![image](https://github.com/oppia/oppia/assets/119884665/5de845f6-a31f-40d7-8ee0-593ecf25bf40)

Though the tests for updating  audio language  and updating site language are similar, tests for 'audio' have passed, but not for 'site' due to click interceptions.

I have immediately re-tested locally, it passed.

[Screencast from 21-01-24 10:52:51 PM IST.webm](https://github.com/oppia/oppia/assets/119884665/25d2431b-240f-44e0-9165-65d57bcd49f4)

Thanks!

seanlip commented:
@AkashPaloju I re-ran, but I'm not sure this is a flake since (a) I've never seen it before and (b) it seems related to your changes. Please try running locally several times (try using the exact same command that CI uses). If it fails at least once, it's important to get to the bottom of it and fix it here so it doesn't cause problems for other people in develop.

AkashPaloju committed:
Fixed a navigation e2e test

AkashPaloju commented:
@seanlip All tests are passed now. PTAL

seanlip reviewed (changes requested):
Thanks @AkashPaloju. Some notes from the PM side:

- The right edge of the "Save" button should align with the right edge of the textboxes, input fields, etc. Could you please fix this?
- Could you please show a video for what happens when the user enters some stuff partially and then navigates elsewhere or closes the page? We should have a confirmation popup that alerts them that they might lose their changes (and ideally alerts them to save first, but not sure if this is possible).

Thanks!

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -1191,6 +1190,7 @@
 	"I18N_PREFERENCES_PROFILE_PICTURE_DRAG": "Text displayed in the preferences page. - Text of the dialog shown to upload a profile picture when the user has uploaded an image that is too big and needs to be cropped.",
 	"I18N_PREFERENCES_PROFILE_PICTURE_ERROR": "Text displayed in the preferences page. - Error text of the dialog shown to upload a profile picture. This error is shown when the file uploaded by the user is not an image.",
 	"I18N_PREFERENCES_PROFILE_PICTURE_UPLOAD": "Text displayed in the preferences page. - Title of the dialog shown to upload a profile picture.",
+	"I18N_PREFERENCES_SAVE_CHANGES": "Text displayed in the preferences page. - Text of the Save Changes button.",
```
Translators already know what's in the button -- that's what they're translating, after all. Try to describe this similar to other items here so they can visualize what they are translating.

AkashPaloju replied:
Done

seanlip commented on core/controllers/library_test.py:
```diff
@@ -413,7 +413,10 @@ def test_library_index_handler_for_user_preferred_language(self) -> None:
         self.put_json(
             feconf.PREFERENCES_DATA_URL,
-            {'update_type': 'preferred_language_codes', 'data': ['de']},
+            {'updates': [{
```
Indentation is a bit off. We normally do:

```
        self.put_json(feconf.PREFERENCES_DATA_URL, {
            'updates': [{
                'update_type': ...,
                'data': ['de']
            }]
        })
```
 
Please fix below/elsewhere, too.

AkashPaloju replied:
Done.

seanlip commented on core/controllers/profile.py:
```diff
@@ -334,65 +334,69 @@ def get(self) -> None:
-            # Check if all required keys are in the data dictionary.
-            if not all(key in data for key in required_keys):
+        for update in updates:
+            update_type = update.get('update_type')
```
These are no longer optional parameters, right? If so, should you raise an error if they aren't provided?

AkashPaloju replied:
Done.

seanlip commented on core/controllers/profile.py:
```diff
@@ -334,65 +334,69 @@ def get(self) -> None:
+                # Check if all required keys are in the data dictionary.
+                if not all(key in data for key in required_keys):
+                    raise self.InvalidInputException(
+                        'Expected data to contain the fields,%s, received %s'
```
I'm not sure this is formatted right -- there should probably be a space after the comma.

In any case, this is not that helpful -- could you make the check figure out which keys are missing and only include those in the error message?

AkashPaloju replied:
Done.

seanlip commented on core/controllers/profile.py:
```diff
@@ -334,65 +334,69 @@ def get(self) -> None:
+                        % (required_keys, data))
+                if not all(isinstance(value, bool) for value in data.values()):
+                    raise self.InvalidInputException(
+                        'Expected all values of data to be boolean,received %s'
```
boolean,received --> booleans, received

AkashPaloju replied:
Done

seanlip commented on core/controllers/profile_test.py:
```diff
@@ -67,18 +67,18 @@ def test_page_not_found(self) -> None:
-                '/preferenceshandler/data',
-                {
+                feconf.PREFERENCES_DATA_URL,
+                {'updates': [{
```
See notes on formatting above.

AkashPaloju replied:
Done.

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -65,8 +65,13 @@ export class SubjectInterestsComponent {
 
   validInput(value: string): boolean {
+    // The following regex matches all the special characters and numbers.
+    let invalidRegex = new RegExp(
```
Just wondering, why not allowlist rather than denylist? E.g. do you want emoji in the string?

AkashPaloju replied:
Done

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -65,8 +65,13 @@ export class SubjectInterestsComponent {
     return value === value.toLowerCase() &&
-      this.subjectInterests.indexOf(value) < 0 ? true : false;
+      this.subjectInterests.indexOf(value) < 0 &&
+      invalidRegex.test(value) === false ? true : false;
```
You should be able to just use a boolean cast for this part, rather than `?:`. E.g. `!invalidRegex.test(value)` would probably do.

AkashPaloju replied:
Done

seanlip commented on core/templates/pages/preferences-page/preferences-page.component.spec.ts:
```diff
@@ -276,69 +284,64 @@ describe('Preferences Page Component', () => {
       }));
 
-      it('should save preferred site language codes', fakeAsync(() => {
+      it('should save preferred site languauge code', fakeAsync(() => {
```
Why change this to a misspelling?

AkashPaloju replied:
I observed ,  'preferred site language code'  is a single string and site language  cannot be  2 or more languages at a time . And  'preferred language codes'  is an array of strings, we use it for checking if an exploration is available in the user preferred languages. 
Please let me know if I am wrong.
Thanks!

seanlip replied:
That's not what I'm referring to. "language" is misspelled in your updated version.

AkashPaloju replied:
Sorry , for the silly mistake. Done.

seanlip commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -85,6 +84,20 @@ export class PreferencesPageComponent {
+  newProfilePictureDataUrl!: string;
+  pageShouldBeReloaded: boolean = false;
+  hasUnsavedChanges = false;
+  unsavedChanges: { [key: string]: boolean } = {
```
Add a comment explaining what this represents.

After the data is saved then does this get reset? (Ideally it would be nice to avoid a full page reload, though I guess we might still need it if e.g. the default language changes. But we should probably restrict that only to the cases where it is actually needed.)

AkashPaloju replied:
Yes, this get resets at the end '.then' block in  a for loop in savePreferences function below. 
The 'pageShouldBeReloaded' is set to true when we need to update profile image only, and after page reloads  all these variables, unsavedChanges, pageShouldBeReloaded,hasUnsavedChanges,  will have false value by default.

seanlip replied:
OK, sounds fine, thanks.

AkashPaloju replied:
done

seanlip commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -110,68 +123,51 @@ export class PreferencesPageComponent {
-  ): void {
+  handleBioChange(): void {
+    this.hasUnsavedChanges = true;
+    this.unsavedChanges.user_bio = true;
```
This doesn't make sense to me. Why are we setting user_bio to true?

Also I am not keen on naming functions with "handle" -- that doesn't convey adequately what the function does. Try to use more specific names here. 

I'm guessing your approach is something like, track which fields are modified and set them to true if they are "dirty". In which case, maybe make hasUnsavedChanges a function that returns true if at least one value in this.unsavedChanges is true.  And maybe this.unsavedChanges could be something like this.fieldsWithChangedValues since it doesn't actually store the changes themselves.

In general -- naming is important! Try to name things properly, otherwise you will confuse future devs.

AkashPaloju replied:
Here we are setting unsavedChnages.'user_bio' to true, so that we can filter out the update_type in savePreferences function below.
I will change the names as specified.

seanlip commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -331,11 +315,86 @@ export class PreferencesPageComponent {
+      { update_type: 'email_preferences', data: emailData }
+    ];
+    updates = updates.filter((update) => {
+      return this.unsavedChanges[update.update_type] === true;
```
Why not instead iterate through this.unsavedChanges and, as items are true, add them to the updates list? Then if new fields are added, there won't be a chance of the dev forgetting to specify them in this local constant as well -- everything draws from unsavedChanges as the source of truth.

AkashPaloju replied:
If we iterate through this.unsavedChanges, we can know only which 'update_type' we need to post to server , we  won't  have the  new data needs to be updated  in a loop, 'update_type' are in snake case and the data in the page is in camelCase variables. Eg. `['updates':{ 'update_type': 'user_bio','data':this.userBio}]` should be posted to backend and `this.userBio` or `this['userBio']`will have the new user bio.
  If we want to proceed as you said, then 
  - we have to do it with 8 if clauses (even then, when a new field is added, we need to add 1 more if-clause)
 OR
- make 'updates' a global object array which contains updates objects  we need to post to server, then we can push or change this object array for each user input. But we need to update this array more frequently (Now we are making a copy only when Save changes button is clicked).
OR
- this also works fine, but I don't know this is recommended or not.
```
// A function which returns the camelCase version of the update_type.
private _camelCaseOfUpdateType (updateType: string): string { XXX }
...
let updates: UpdatePreferenceDict[] = [];
for (let updateType in this.fieldsWithChangedValues) {
  if (this.fieldsWithChangedValues[updateType]) {
    updates.push({
      update_type: updateType,
      data: this[ _camelCaseOfUpdateType(updateType) ]
    });
  }
}
```
Please let me know your suggestions on this.
Thanks!

seanlip replied:
Create a UserPreferences object model class to represent a user's preferences. It should have fields like defaultDashboard, profilePictureDataUrl, etc.

Update this component to create a UserPreferences model from the backend data, and update it as needed. Don't store local variables in the component for all the different fields. Include "dirty" tracking as private variables/methods inside the model class. Finally, expose a function in that class along the lines of getUpdatedFields(), and use that to generate the relevant thing to send to the backend on-save. (You can also create a strictly-typed constant dict near the top that maps the model field names to the backend update_types.)

AkashPaloju replied:
Done.

seanlip commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -331,11 +315,86 @@ export class PreferencesPageComponent {
+    });
+
+    if (AssetsBackendApiService.EMULATOR_MODE) {
+      // Remove 'profile_picture_data_url' from updates if the emulator mode is
```
Did we do this before? If not we shouldn't do it now. The behaviour should be preserved.

Also I'm not sure how valid this is. Why can't we do this in dev mode? We can upload images to explorations in dev mode AFAIK, why can't we do it here?

AkashPaloju replied:
![Screenshot from 2024-01-25 23-45-11](https://github.com/oppia/oppia/assets/119884665/7fea8436-79c3-4514-bdf4-ea4ffaadb049)
As seen in the earlier code in the above image, if user is in emulator mode the file is saved to local storage, else it is posted to backend server.

seanlip replied:
Oh, I see. But could you check why it's not actually possible to save in the backend in emulator mode? Offhand I don't see a reason why we can't do that, and if we can get rid of this branching then the code will be easier to maintain.

seanlip replied:
@AkashPaloju You haven't replied to this comment.

AkashPaloju replied:
I have tried uploading profile image after removing 'if'-branches but it not worked. I have checked the code thoroughly in exploration editor handlers, imageUploadHandler, file_services, storage_services etc.. and this profile picture logic in backend but could not find the reason till now and investigating further.

I guess the developer who crafted this code had a compelling rationale behind incorporating this 'if' branching.

seanlip replied:
Can you find the PR which introduced this? We should not assume -- if you don't know the "compelling rationale", let's find it out.

AkashPaloju replied:
I guess it would be better if we open a new issue for this as this PR has not introduced this problem. What do you say?

AkashPaloju replied:
> Can you find the PR which introduced this? We should not assume -- if you don't know the "compelling rationale", let's find it out.

hi @seanlip. This [PR#16521](https://github.com/oppia/oppia/pull/16521/files#diff-0178d27d866bd962c1b96dd07b0e3a0cc8216714a3c8344cea73c3e4853ced38)  introduced this. But after reading most of the PR discussions(there are almost 350 discussions and 150 commits!) I could not get the exact reason. This PR is not linked to any issue may be a GSoC project milestone.

Hello, @lkbhitesh07  Could you please let us know why the profile-image uploads to local storage instead of GCS here , unlike in exploration creator page, if you still remember?
Thanks!

lkbhitesh07 replied:
Hey @AkashPaloju, You can always assign the person you want to take the next action otherwise that person might not get notified.
Regarding your question, I do not exactly remember why we went ahead with storing the images in the local storage. Me and my mentor were having some discussion around it and we agreed upon saving the images in local storage for the DEV mode. Now come to think of it we could've used the `dev_mode_storage_services.py` file to store the data in the backend, I think it would've worked.
Thanks

seanlip replied:
@AkashPaloju Thanks for investigating. Based on Hitesh's response, I think it's fine to file an issue for this, and put a TODO in the code pointing to that issue. That would suffice -- thanks!

AkashPaloju replied:
Done. Added TODO comment and filed issue #19737

seanlip commented on core/templates/services/context.service.spec.ts:
```diff
@@ -180,7 +180,7 @@ describe('Context service', () => {
   describe('behavior in the topic editor view', () => {
     beforeEach(() => {
-      ecs = TestBed.get(ContextService);
+      ecs = TestBed.inject(ContextService);
```
Out of curiosity, why did this change? (I don't mind, just wondering.)

AkashPaloju replied:
Got deprecated error for this when I run full e2e tests locally, I don't know why, Should I undo this?

seanlip replied:
What's the full stacktrace of the error?

What do we do in other files?

AkashPaloju replied:
It is resolved  for me now. That error is not visible in CI tests here though.
In some files , 'inject' is used but in older files 'get' is used.

seanlip replied:
OK. I'm fine with following the newer convention if it works.

AkashPaloju replied:
Ok

seanlip commented on core/templates/services/user-backend-api.service.ts:
```diff
@@ -64,6 +64,11 @@ interface LoginUrlResponseDict {
   'login_url': string;
 }
 
+export interface UpdatePreferenceDict {
```
Is it possible to make this typing stricter, with the `data` type determined based on the value of `update_type`?

AkashPaloju replied:
Done.

seanlip commented on core/tests/webdriverio_utils/PreferencesPage.js:
```diff
@@ -82,44 +92,54 @@ var PreferencesPage = function() {
 
   this.editUserBio = async function(bio) {
     await action.addValue('User bio field', userBioElement, bio);
+    // Click on a neutral element to make User bio element out of focus.
```
Why is this needed?

AkashPaloju replied:
Observe the Save Changes button's disability status and console statements which are logged when (change) event is triggered for textarea in the following video
[Screencast from 26-01-24 07:20:01 PM IST.webm](https://github.com/oppia/oppia/assets/119884665/2df31671-44cc-4ea7-a99c-bd0fc91245ca)


The below problem comes when we don't use it, beacuse the textarea is not out of focus means the (change) event is not triggered, so 'hasUnsavedChanges' is still false and 'Save changes' button is disabled.
[Screencast1](https://github.com/oppia/oppia/assets/119884665/6c48e71c-8986-4b5c-a905-fadded85de53)

[Screencast from 26-01-24 06:14:28 PM IST.webm](https://github.com/oppia/oppia/assets/119884665/b04c8fd7-e459-470b-8f88-097b56a756e0)

AkashPaloju replied:
Moved the logic to 'saveNewChanges' so this is not needed and removed now.

seanlip commented on core/tests/webdriverio_utils/PreferencesPage.js:
```diff
@@ -82,44 +92,54 @@ var PreferencesPage = function() {
     await action.clear(inputFieldName, userBioElement);
-    await saveNewChanges('User Bio');
     await action.setValue(inputFieldName, userBioElement, bio);
+    // Click on a neutral element to make User bio element out of focus.
```
Why is this needed? Also how is setUserBio different from editUserBio?

AkashPaloju replied:
'setUserBio' replaces the old value with the given value.
'editUserBio' just appends the given string to the older bio.

seanlip replied:
OK. I think it's fine to keep as-is, please just add JSDoc if that's OK so that future devs are not confused.

AkashPaloju replied:
done

seanlip replied:
Not done. Please don't say something is done when it isn't.

AkashPaloju replied:
Sorry for slight confusion with the above comment thread near 'editUserBio' function --- I thought you have asked JSDoc for this lines 
'// Click on a neutral element to make User bio element out of focus'
'await action.click('Preferences page header', pageHeader);' as you have commented on this lines. So I have put 'Done' as I have removed those 2 lines to 'saveNewChanges' function.

Got it now. Will make a comment explaining what is 'setUserBio' and 'editUserBio'

seanlip reviewed (changes requested):
Also, took a look and left code comments inline!

AkashPaloju committed:
Addressing review comments 1

DubeySandeep commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -331,11 +315,86 @@ export class PreferencesPageComponent {
   }
+
+  savePreferences(): void {
+    if (!this.hasUnsavedChanges) {
```
Maybe use  (reactive-form)`FormGroup` and `FormControl` for the preference page to avoid so many state management codes. 

https://angular.io/guide/reactive-forms

AkashPaloju replied:
hello @DubeySandeep  I have made the functionality using a User Preferences model as suggested by Sean few days ago. PTAL at the changes. and if you still suggest FormGroup is needed here, I will do that.
Thanks!

DubeySandeep replied:
@AkashPaloju Do you think having FormGroup will reduce the complexity of this component? If yes, we can do it using FormGroup!

AkashPaloju replied:
@DubeySandeep sorry, busy due to exams, will reply in a day or 2.

AkashPaloju replied:
Yes, It will reduce the complexity of this page, but I need to check, is it possible to make everything(especially child components) as a form element . Working on it. Will update very soon.

AkashPaloju replied:
@DubeySandeep PTAL, Thanks for this idea.

AkashPaloju committed:
Adjusted save button location

AkashPaloju committed:
Removed a unused Email preferences variable

AkashPaloju committed:
Fixed a failing frontend coverage test

AkashPaloju commented:
@seanlip PTAL the changes. Except a flaky e2e test in topicAndStoryViewer suite remaining all tests are passed.

seanlip commented on core/controllers/profile.py:
```diff
@@ -354,15 +354,21 @@ def put(self) -> None:
-                        'Expected data to contain the fields,%s, received %s'
-                        % (required_keys, data))
-                if not all(isinstance(value, bool) for value in data.values()):
+                        'Expected data dict to have all the required keys.'
```
Add a space at the end of the string, otherwise you will get "keys.Missing" in the string.

AkashPaloju replied:
Done.

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -65,13 +65,11 @@ export class SubjectInterestsComponent {
-      '[~`!@#$%^&*()_+={[}\\]|\\:;"\'<,>.?/\\-0-9]');
+    // The following regex matches only lowercase
+    // alphabetic characters and spaces.
+    let validRegex = new RegExp('^[a-z\\s]*$');
```
Do we / should we have similar validation in the backend?

AkashPaloju replied:
there was a similar validation in the backend which I have removed in this [PR](https://github.com/oppia/oppia/pull/19633/files) as suggested by you.

![image](https://github.com/oppia/oppia/assets/119884665/19a174e7-0978-425c-afeb-1e1ad9bc59de)

![image](https://github.com/oppia/oppia/assets/119884665/566e72e9-3db7-45f3-8b47-8331ce6ab73e) in [constants.ts](https://github.com/oppia/oppia/blob/develop/assets/constants.ts)

seanlip replied:
OK, thanks, that's seems fine then.

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -65,13 +65,11 @@ export class SubjectInterestsComponent {
-    return value === value.toLowerCase() &&
-      this.subjectInterests.indexOf(value) < 0 &&
-      invalidRegex.test(value) === false ? true : false;
+    return validRegex.test(value) && this.subjectInterests.indexOf(value) < 0;
```
Maybe use .includes() or "=== -1" -- otherwise it takes a bit of reasoning to work out what this condition means.

AkashPaloju replied:
Done

seanlip commented on core/tests/webdriverio_utils/PreferencesPage.js:
```diff
@@ -49,6 +49,8 @@ var PreferencesPage = function() {
   var userInterestsInput = $('.e2e-test-subject-interests-input');
 
   var saveNewChanges = async function(fieldName) {
+    await action.click('Navbar Button', navBar);
```
What is the "navbar button" and why do we need to click it?

If this is the "neutral element" thing then you should explain so that devs know what is going on and what happens if it is removed.

AkashPaloju replied:
Done.

seanlip commented:
Assigning @DubeySandeep and @kevintab95 for codeowner reviews, too.

AkashPaloju committed:
Address review comments 2

AkashPaloju commented:
PTAL at the changes. Thanks!

seanlip reviewed (commented):
Thanks @AkashPaloju -- it looks good, just one small comment here. Thanks! https://github.com/oppia/oppia/pull/19529#discussion_r1486456555

AkashPaloju committed:
Added TODO comment for issue 19737

seanlip reviewed (approved):
Thanks, LGTM for my codeowner files.

seanlip commented:
@AkashPaloju Are you planning to reply to @DubeySandeep's comment? https://github.com/oppia/oppia/pull/19529#discussion_r1486645620

lkbhitesh07 reviewed (approved):
LGTM! for the code owner files. I have one question @AkashPaloju, In the proof-of-work video, where we are uploading the profile picture, is it possible to showcase the updated profile-picture just after uploading it and not after saving it. Our user might think that the profile picture is not uploaded correctly.
cc/- @seanlip 
Thanks

seanlip reviewed (changes requested):
Oh, that is a great point re the image, @lkbhitesh07. I totally missed that, oops.

@AkashPaloju could you please fix that? As Hitesh says, it is not a great user experience, it feels like the user's image upload did not succeed. (Another way to look at it is, all other edits to the preferences page get applied immediately -- at least temporarily, anyway -- so, why not this one?)

DubeySandeep commented:
@AkashPaloju Any update on [this thread](https://github.com/oppia/oppia/pull/19529#discussion_r1486645620)?

AkashPaloju committed:
Added profile Image preview after uploading

AkashPaloju commented:
> Oh, that is a great point re the image, @lkbhitesh07. I totally missed that, oops.
> 
> @AkashPaloju could you please fix that? As Hitesh says, it is not a great user experience, it feels like the user's image upload did not succeed. (Another way to look at it is, all other edits to the preferences page get applied immediately -- at least temporarily, anyway -- so, why not this one?)

Done, updated the video also.
Thanks!

AkashPaloju committed:
Use reactive form in preferences page

AkashPaloju committed:
Remove an unused function

AkashPaloju committed:
fixing coverage test

AkashPaloju committed:
updating the comments

AkashPaloju committed:
fixed a e2e test

AkashPaloju commented:
> LGTM! for the code owner files. I have one question @AkashPaloju, In the proof-of-work video, where we are uploading the profile picture, is it possible to showcase the updated profile-picture just after uploading it and not after saving it. Our user might think that the profile picture is not uploaded correctly. cc/- @seanlip Thanks

@lkbhitesh07 PTAL, updated video in the proof of work also.

Thanks for catching this and for your clarification here : )

seanlip reviewed (changes requested):
Thanks @AkashPaloju. Only one comment from me: I don't see a video showing what happens if the user changes some fields and leaves the page -- an "are you sure" confirmation dialog should appear in such cases. Could you demonstrate this on both web and mobile? Otherwise we may get complaints from users who spend a bunch of time updating their preferences and then lose their work.

Thanks!

AkashPaloju commented:
> Thanks @AkashPaloju. Only one comment from me: I don't see a video showing what happens if the user changes some fields and leaves the page -- an "are you sure" confirmation dialog should appear in such cases. Could you demonstrate this on both web and mobile? Otherwise we may get complaints from users who spend a bunch of time updating their preferences and then lose their work.
> 
> Thanks!

I have included that in the image preview video at last. If a user uploads a video and doesn't save it there will be a confirmation popup.

seanlip commented:
Thanks @AkashPaloju! One note -- in the newest mobile video you added, it seems that, when the user scrolls up, the bottom bar disappears. Shouldn't it be always in view? Otherwise you can run into a situation where the user gets the scary warning but isn't aware of what to do next.

AkashPaloju commented:
> Thanks @AkashPaloju! One note -- in the newest mobile video you added, it seems that, when the user scrolls up, the bottom bar disappears. Shouldn't it be always in view? Otherwise you can run into a situation where the user gets the "be careful!" warning but isn't aware of what to do next.

@seanlip Actually it is due to scrolling of the browser  , it appeared as gone away. But it is actually present in the mobile screen at the bottom. PTAL at this video I have tried to explain in it. If we zoom out we can see that it is still at the bottom. Updated at the PR description too.

[latest-video.webm](https://github.com/oppia/oppia/assets/119884665/035b5410-2c7f-49b5-b93d-63e4ba1e3979)

seanlip reviewed (approved):
Ah sorry @AkashPaloju, you're right, I missed that! LGTM then :) Thanks for making the fixes!

AkashPaloju committed:
Fixing button interceptions in a e2e test

AkashPaloju commented:
PTAL  at this video this test failed in the merge-queue . Fixed that. This test was failed due to button interceptions so, clicking a nuetral button in between will resolve this.

[screencast1.webm](https://github.com/oppia/oppia/assets/119884665/08992554-161c-4b54-88e0-8ea906b9c08c)

[screencast-after-fix.webm](https://github.com/oppia/oppia/assets/119884665/82b50c6c-6f73-4d87-977b-0894c744a14f)

AkashPaloju committed:
Added checks before click in failing e2e test

AkashPaloju commented:
@seanlip PTAL I fixed the  failed  test which was failed in the merge-queue . This test was failed due to button interceptions  which was caused due to non visibility of the required button. so, added checks for visibility and clickablility. I have run the e2e test 5 times locally after adding those checks, it passed 5 times.



A flake e2e test which is not at related to my changes is failing in coreEditor page.
Thanks!

lkbhitesh07 commented:
> > LGTM! for the code owner files. I have one question @AkashPaloju, In the proof-of-work video, where we are uploading the profile picture, is it possible to showcase the updated profile-picture just after uploading it and not after saving it. Our user might think that the profile picture is not uploaded correctly. cc/- @seanlip Thanks
> 
> @lkbhitesh07 PTAL, updated video in the proof of work also.
> 
> Thanks for catching this and for your clarification [here](https://github.com/oppia/oppia/pull/19529#discussion_r1485011410) : )

Thanks @AkashPaloju, LGTM!

seanlip reviewed (changes requested):
@AkashPaloju An e2e test is failing in the merge queue. You might want to take a look at it, it relates to the preferences page.
![Screenshot from 2024-02-26 02-52-05](https://github.com/oppia/oppia/assets/10575562/c8ffc56e-b4c4-4290-bec2-eb71e9797b91)

AkashPaloju committed:
Fixed e2e test by scrolling into view

AkashPaloju commented:
@seanlip PTAL , The tests are passed now
Thanks!

AkashPaloju committed:
Generalise the e2e test scrolling

AkashPaloju commented:
@seanlip I have  generalized the scrolling PTAL.

Why this error occured:

If you observe, see the above image ( which you have attached ) the element ('English' in the list) which should be clicked is not Displayed in Viewport , WebdriverIO couldn't correctly scroll the element into the viewport like earlier because of the sticky nature of the 'save changes' button added in this PR(which was the only UI change made in the PR). I tested by removing the sticky nature of the button it passed all the times. So, to tackle it we need to scroll it until it is visible in the viewport and then click the dropdown.
Update the same in the [debugging doc](https://docs.google.com/document/d/1r62BDuRS6MmeJ4blZhtuBb69StqKbrx5J912QyITsA4/edit?usp=sharing)

Thanks!

seanlip commented:
@AkashPaloju Thanks -- I need a little more detail, and I think you'll need to add that detail as code comments too so that future developers know what's going on.

Specifically: you say "WebdriverIO couldn't correctly scroll the element into the viewport like earlier because of the sticky nature of the 'save changes' button" I get that the functionality is different from earlier, but I don't understand why WebdriverIO can't correctly scroll the element into the viewport. Can you explain why (without reference to the earlier version of the page)? I.e. explain it in terms of the current page, imagining that there wasn't an earlier version of the page and this was the first version that was ever implemented.

AkashPaloju commented:
@seanlip

Apologies, the error is not due to the sticky nature of the save button, actually the error is not due to any changes made in the PR.

_Here is a detailed analysis of the error:_

_Please re-read the [debugging doc](https://docs.google.com/document/d/1r62BDuRS6MmeJ4blZhtuBb69StqKbrx5J912QyITsA4/edit?usp=sharing) for the context of the error._

Take a look at the screenshot when the test failed.

Explanation:

After reading the error log (attached in debugging doc) and inspecting the preferences page [elements.webm](https://github.com/oppia/oppia/assets/119884665/f999a03b-275b-4482-abe5-64892a38b164)

_Why the test failed?_
The test failed because when it tried to click the 'English' option, the 'Search' option received the click.

_Why the 'Search' option received the click_
The 'click' method of webdriverIO as mentioned [here](http://v4.webdriver.io/api/action/click.html) fails to click the underlying element ( i.e. the 'English' option)

_How do we know that 'English' is under the 'search' option_

Please check the "Known Problems" section here of the ngx-mat-select-search . [link](https://www.npmjs.com/package/ngx-mat-select-search#known-problems)
which was used in the  [preferences-page/form-fields/preferred-language-selector.component.html](https://github.com/oppia/oppia/blob/develop/core/templates/pages/preferences-page/form-fields/preferred-language-selector.component.html)

Also  please look at this screencast [site-lang.webm](https://github.com/oppia/oppia/assets/119884665/82cd9db4-dbd4-4c2a-bebe-a5facb58ce83). In this I have scrolled up the items from 'Espanol' such that the 'English' option is in viewport (very much little part is in viewport about only 1px). Then I have clicked the narrow area (between the  'Portugues' and 'Search' ), so then the 'English' was selected. 
But webdriverIO is sometimes failing to click the narrow area. 
The 'click' method scrolls the element by little amount until it comes into the viewport (if it is not present in viewport as said in the above 'click'  documentation).

In the example  screencast (in which the position of the all dropdown options are same as in the screenshot of the failed test) I have shown above , the element is in viewport but the probability of correctly clicking the 'English' option depends solely on webdriverIO. If the click is not received by 'English' it will be received by 'Search' option.

So the tests are failing sometimes and passing other times.

So we need a generalized programmatic scrolling to fix this as recomendded in the same 'click' documentation. 

As the explanation is somewhat lengthy, I'm thinking to put this 'comment' link as a code comment. What do you say? 

Thanks!

seanlip commented:
Hi @AkashPaloju, thanks, seems reasonable, and the webdriverio doc helps.

Code comments should be self-contained. Please summarize this clearly with links to external authoritative references (like the webdriverio and ngx-mat-select-search links, but not this GitHub PR) and leave that as a code comment. It is fine if the code comment is lengthy; the main requirement is that it must be clear (please check with a fellow developer and get feedback if you're not sure). You can assign this for review once you are happy with the code to be submitted (including the comment mentioned here).

AkashPaloju committed:
Added comment explaining the need of scrolling

AkashPaloju committed:
Merge remote-tracking branch 'upstream/develop' into sticky-save

AkashPaloju committed:
Applied prettier formatter as per new update

seanlip reviewed (approved):
Looks good. Thanks @AkashPaloju!
